aspnetcore/security/authentication/claims.md

```diff
@@ -0,0 +1,155 @@
+    })
+   .AddCookie()
+   .AddOpenIdConnect(options =>
+   {
+       options.SignInScheme = "Cookies";
+       options.Authority = "-your-identity-provider-";
+       options.RequireHttpsMetadata = true;
+       options.ClientId = "-your-clientid-";
+       options.ClientSecret = "-your-client-secret-from-user-secrets-or-keyvault";
+       options.ResponseType = "code";
+       options.UsePkce = true;
+       options.Scope.Add("profile");
+       options.SaveTokens = true;
+       options.GetClaimsFromUserInfoEndpoint = true;
+       options.ClaimActions.MapUniqueJsonKey("preferred_username", "preferred_username");
+       options.ClaimActions.MapUniqueJsonKey("gender", "gender");
+   }); 
+```
+
+## Name claim and role claim mapping
+
+The **Name** claim and the **Role** claim are mapped to default properties in the ASP.NET Core HTTP context. Sometimes it is required to use different claims for the default properties, or the name claim and the role claim do not match the default values. The claims can be mapped using the **TokenValidationParameters** property and set to any claim as required. The values from the claims can be used directly in the HttpContext **User.Identity.Name** property and the roles.
+
+If the **User.Identity.Name** has no value or the roles are missing, please check the values in the returned claims and set the **NameClaimType** and the **RoleClaimType** values. The returned claims from the client authentication can be viewed in the HTTP context.
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services.AddAuthentication(options =>
+    {
+        options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+        options.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
+    })
+   .AddCookie()
+   .AddOpenIdConnect(options =>
+   {
+       // other options...
+       options.TokenValidationParameters = new TokenValidationParameters
+       {
+         NameClaimType = "email", 
+         // RoleClaimType = "role"
+       };
+   });
+```
+
+## Claims namespaces, default namespaces
+
+ASP.NET Core adds default namespaces to some known claims which might not be required in your application. You can disable this namespace adding, if you would like to use the claims exactly like the Open ID Connect server created. 
+
+```csharp
+public void Configure(IApplicationBuilder app)
+{
+    JwtSecurityTokenHandler.DefaultInboundClaimTypeMap.Clear();
+```
+
+## Extending or adding custom claims using IClaimsTransformation
+
+The **IClaimsTransformation** interface can be used to add extra claims to the **ClaimsPrincipal** class. The interface requires a single method **TransformAsync**. This method might get called multiple times. Only add a new claim if it does not already exist in the **ClaimsPrincipal**. A **ClaimsIdentity** is created to add the new claims and this can be added to the **ClaimsPrincipal**.
```

Comment:
```suggestion
The `IClaimsTransformation` interface can be used to add extra claims to the `ClaimsPrincipal` class. The interface requires a single method `TransformAsync`. This method might get called multiple times. Only add a new claim if it does not already exist in the `ClaimsPrincipal`. A `ClaimsIdentity` is created to add the new claims and this can be added to the `ClaimsPrincipal`.
```

Reply:
fixed